quinn/src/recv_stream.rs

```diff
@@ -21,8 +21,40 @@ use crate::{
 /// - A variant of [`ReadError`] has been yielded by a read call
 /// - [`stop()`] was called explicitly
 ///
+/// # Closing a stream
+///
+/// When a stream is expected to be closed gracefully the sender should call
+/// [`SendStream::finish`].  However there is no guarantee the connected [`RecvStream`] will
+/// receive this End Of File (EOF) indicator in the same QUIC frame as the last frame which
+/// carried data.
+///
+/// So even if the application layer logic already knows it read all the data because it
+/// does its own framing, it must still read from the [`RecvStream`] until EOF.  Otherwise
+/// it risks inadvertently calling [`RecvStream::stop`] if it drops the stream.  And calling
+/// [`RecvStream::stop`] could result in the connected [`SendStream::finish`] call failing
+/// with a [`WriteError::Stopped`] error.
+///
+/// For example if exactly 10 bytes are to be read, you still need to explicitly read EOF:
+///
+/// ```ignore
+/// // In the sending task
+/// send_stream.write(&b"0123456789"[..]).await?;
+/// send_stream.finish()?;
+///
+/// // In the receiving task
+/// let mut buf = [0u8; 10];
+/// let data = recv_stream.read_exact(&mut buf);
+/// if let Some(chunk) = recv_stream.read_chunk(8, false).await? {
+///     reader.stop(0u8.into()).ok();
+/// }
```

Comment:
This is a little more terse, and has exactly the same effect if the `recv_stream` is dropped after `?` bails out:
```suggestion
/// recv_stream.read_to_end(0).await?;
```

Reply:
I did consider that option, but preferred the `.read_chunk()` approach:

- The amount of data being read is unlimited and allocated in a `Vec` when using `.read_to_end()`.  The receiver has no control over this.
- `.read_chunk(n, false)` has the benefit there won't even be any delays due to reordering or retransmission.
- Returning from the function may or may not drop the reader, that depends on the context of the caller.  For an example I prefer the explicit call to `.stop()`.

Reply:
Huh, sorry I did misread `read_to_end`.  Makes sense, I'd still leave in the explicit stop though for the example.

Reply:
I suppose it does make the intent significantly clearer; fine by me.